File: src/main/java/dev/slimevr/vr/processor/TransformNode.java
package dev.slimevr.vr.processor;

import com.jme3.math.Transform;
import io.eiren.util.collections.FastList;

import java.util.List;
import java.util.function.Consumer;


public class TransformNode {

	public final Transform localTransform = new Transform();
	public final Transform worldTransform = new Transform();
	public final List<TransformNode> children = new FastList<>();
	public boolean localRotation = false;
	protected TransformNode parent;
	protected String name;

	public TransformNode(String name, boolean localRotation) {
		this.name = name;
		this.localRotation = localRotation;
	}

	public TransformNode(String name) {
		this(name, true);
	}

	public void attachChild(TransformNode node) {
		if (node.parent != null) {
			throw new IllegalArgumentException("The child node must not already have a parent");
		}

		this.children.add(node);
		node.parent = this;
	}

	public TransformNode getParent() {
		return parent;
	}

	public void update() {
		updateWorldTransforms(); // Call update on each frame because we have
									// relatively few nodes
		for (TransformNode node : children) {
			node.update();
		}
	}

	protected synchronized void updateWorldTransforms() {
		if (parent == null) {
			worldTransform.set(localTransform);
		} else {
			worldTransform.set(localTransform);
			if (localRotation)
				worldTransform.combineWithParent(parent.worldTransform);
			else
				combineWithParentGlobalRotation(parent.worldTransform);
		}
	}

	public void depthFirstTraversal(Consumer<TransformNode> visitor) {
		for (TransformNode node : children) {
			node.depthFirstTraversal(visitor);
		}
		visitor.accept(this);
	}

	public String getName() {
		return name;
	}

	public void combineWithParentGlobalRotation(Transform parent) {
		worldTransform.getScale().multLocal(parent.getScale());
		worldTransform.getTranslation().multLocal(parent.getScale());

		parent
			.getRotation()
			.multLocal(worldTransform.getTranslation())
			.addLocal(parent.getTranslation());
	}
}

File: server/core/src/main/java/dev/slimevr/tracking/processor/TransformNode.kt
package dev.slimevr.tracking.processor

import io.eiren.util.ann.ThreadSafe
import io.github.axisangles.ktmath.Transform
import java.util.concurrent.CopyOnWriteArrayList
import java.util.function.Consumer

/**
 * Represents a joint
 */
class TransformNode(val localRotation: Boolean) {
	val localTransform = Transform()
	val worldTransform = Transform()
	var parent: TransformNode? = null
		private set
	val children: MutableList<TransformNode> = CopyOnWriteArrayList()

	fun attachChild(node: TransformNode) {
		require(node.parent == null) { "The child node must not already have a parent." }
		children.add(node)
		node.parent = this
	}

	@ThreadSafe
	fun update() {
		// Update transform
		updateWorldTransforms()

		// Update children
		for (node in children) {
			node.update()
		}
	}

	@ThreadSafe
	fun updateThisNode() {
		updateWorldTransforms()
	}

	@Synchronized
	private fun updateWorldTransforms() {
		worldTransform.set(localTransform)
		parent?.let {
			if (localRotation) {
				worldTransform.combineWithParent(it.worldTransform)
			} else {
				combineWithParentGlobalRotation(it.worldTransform)
			}
		}
	}

	fun depthFirstTraversal(visitor: Consumer<TransformNode?>) {
		for (node in children) {
			node.depthFirstTraversal(visitor)
		}
		visitor.accept(this)
	}

	private fun combineWithParentGlobalRotation(parent: Transform) {
		worldTransform.scale = worldTransform.scale hadamard parent.scale
		val scaledTranslation = worldTransform.translation hadamard parent.scale
		worldTransform.translation = parent.rotation.sandwich(scaledTranslation) + parent.translation
	}

	fun detachWithChildren() {
		for (child in children) {
			child.parent = null
		}
		children.clear()
		parent?.children?.remove(this)
		parent = null
	}
}
